### DISCUSSION TITLE
[4/22] Neural Architecture Search (Part I)

### DISCUSSION THREAD

**Discussion by tjdtnsu** [Pseudo-Lab/EfficientML#14] [General]
.

**Comment by namgyu-youn** [comment]
[ ~ p.42]
### MACs (Recap)
- MACs is an indirect evaluator for efficiency metrics : Add(+) + Multiply(×)
- MACs in Linear Layer : `C_in * C_out` (Simple mapping for input and output neuron)
- [1] DW-ConV is the special case when the number of group is equal to the nember of channels.

### ResNet-50
- Bottleneck architecture (2048D -> 512D -> 2048D)
- Bottleneck (middle-layer) is intended for **spatial** feature extraction.

### ResNeXt
- Unlike ResNet-50, Middle layer (3D ConV) is replaced to groupped ConV
- Key feature, Cardinality is intended for parallel computing. (W, D doesn't always make sense)
- Aggregation block is intended for element wise addition, binary combining operation. (Note this is not same as sum)

### MobileNet
- DW-Conv has some feature following features because of ([1]):
   - Each group operates independently for each channels (**no interaction**)
   - Because of the feature, operation is minimized rather than Groupped ConV
   - Because there is no interaction, related-channel challenges to spatial information learning ... (p)
- (S) Pointwise ConV : By 1D ConV (linear-connection)and weighted sum, MobileNet overcomes above issue

### MobileNet-V2
- [2] DW-ConV is cheaper than Normal ConV because of it has only-one filter.
- Inverted bottleneck is introduced based on ([2])
- Multiple DW-ConV has linear relationship with parameters, not like Activation
- Because Activation has a relationship with the number of feature maps, cost for Activation is challenge.

### ShuffleNet
- Group-ConV has limitation because of its own architecture : separated group (no interaction) ... (p)
- (S) ShuffleNet simply focuses on shuffling, resolving interaction for Group-ConV.

### Transformer - MHSA
- Cost : `O(2 *  (n^2 * d)) (n is length of sequence, d is dimension for embedding)
- Basic cost is `O(N^2)` because each vector traverses for all sentence using inner product.
- Inner product costs as `O(d)`

### Manually designed Neural Network
- SOTA : low MACs, high accuracy
- Many SOTA was found by AutoML, initial approach for NAS (e.g., EfficientNet, Once-for-All)

[p. 43 ~]

### Basics for NAS
- NAS aims to find tiny (efficient) model without compression
- Search space : Candidates deciding which operations to apply (e.g., ConV, DW-ConV, Pooling)

### Cell-level search space
- Network-level vs Cell-level : Once-for-all vs One-by-One
- Reduction Cell : Reduce resolution -> lower spatial information, higher kernel size (down sampling)
- Normal Cell : Iterate searching without modifying I/O format (e.g., kernel size, resolution)

### Controller
- Decision making tool based on RNN and Softmax
- Hidden state decides which value to input (e.g., output of last layer)
- M, N : Candidates for hidden state, Function for two-nodes connection

### Search Space : Topology Connection
- Note. Down sampling has limitation in segmentation task -> Upsampling
- Optimizing down/up sampling is one of the task in NAS
- Compression can be useful with NAS (red-zone -> blue-zone)

### Search Strategy : EfficientNet
- Note. I.A focused on "carefully balancing network depth, width, and resolution."
- EfficientNet introduces "compound scaling" for balance

### Search Strategy : Random & RL
- Random searching can be useful with reward function and non-trivial approach (assigned in homework)
- RL-searching decides values (e.g., width, height), but not practical approach. (maybe cost issue)

### Search Strategy : Gradient Descent
- Super network : All candidates for architecture
- Apply importance : Construct architecture parameter for each operations

### Search Strategy : Evolutionary Search
- Initialization -> Fitness func (??) -> Crossover & Mutation (변이) -> Loop until solution is optimal
- Fitness function will be introduced in next lecture. (assigned in homework)

**Comment by Saebitna** [comment]
**NAS: Automatic technique for designing neural network architectures**

**Primitive operations**
    - (review)  MAC (Multiply-Accumulate): 하나의 곱셈과 그 결과를 누적하는 덧셈을 한 쌍으로 보는 연산으로 딥러닝 레이어에서의 MACs은 출력 하나를 계산하는 데 몇 번의 곱셉덧셈이 필요한가를 세면 됨. 출력 요소 하나 당   가중치 (weight) 개수만큼 곳셈이 필요
    
**Classic building blocks**
    - ResNet50: bottleneck block
    - ResNeXt: grouped convolution
    - MobileNet: depthwise-separable block (depthwise convolution → capture spatial info & 1x1 convolution → fuse/exchange info across different channels)
    - MobileNetV2: inverted bottleneck block       
        → 한계점) 계산량 (MAC)은 비슷하지만 채널 확장 단계에서 중간에 저장 해야하는 activation 크기가 6배로 증가하고, 학습 시 역전파할 때 gradient 계산을 위해서 해당 activation을 모두 저장해야 하므로 메모리 부담이 큼         
    - ShuffleNet: 1x1 group convolution & channel shuffle
    - Transformer: Multi-Head Self-Attention (MHSA)

**Introduction to neural architecture search (NAS)**
- Cell-level search space
    - Normal cell: 이전 2개의 cell의 출력 2개를 입력으로 사용. B개의 블록을 구성 (1번째, 2번째 입력 상태를 선택하고, 입력에 사용할 연산을 선택)하고, 연산 결과 결합 방식을 선택하여 그 결과를 새로운 hidden state 상태로 추가함.
    - Reduction cell: 방식은 Normal cell과 동일한데 Stride=2로 해상도를 절반으로 다운샘플링함.
    - RNN controller로 각 셀의 구성 (인풋 2개, 블록 내부 구성, 결합방식) 을 결정하며 B번 반복함.
- Network-level search space
    - Cell-level에서 찾은 normal/reduction cell을 어떻게 이어붙여서 전체 네트워크를 구성할지를 결정해서 다양한 네트워크 아키텍처를 생성
        - Depth dimension: 각 stage에서 cell을 몇 번 반복할지 결정
        - Resolution dimension: 인풋 이미지 크기 (해상도) 결정
        - Width dimension: 각 stage에서 cell이 출력할 채널 수 결정
        - Kernel dimension: depthwise convolution을 사용하는 모델의 경우 각 convolution마다 kernel size를 결정
        - Topology connention: 각 레이어마다 donsampling (stride)을 결정
       
- Search Strategy
    - Grid search
    - Random search
    - Reinforcement search
    - Gradient descent → DART (연산 선택을 softmax 혼합으로 완화 → 연속적 매개변수로 gradient descent)
    - Evolutionary search

**Comment by surisurikim** [comment]
- 전통적인 수작업 모델 설계는 사람의 경험과 시행착오에 의존하는 방식이었지만, 
- NAS는 GPU 등 머신 리소스를 활용하여 수많은 구조를 평가하고 최적의 구조를 찾는다. 
- 이로써 하드웨어 제약까지 고려한 맞춤형 모델 설계가 가능해짐.
---
- NAS의 3가지 구성 요소
    1. Search Space : 탐색할 수 있는 아키텍처들의 집합 (가능한 설계 후보들)
        - Cell-level Search Space (셀 단위 탐색)
            - Normal Cell, Reduction Cell
            - 전체를 다 탐색하는 건 비현실적 -> 효율적인 탐색 알고리즘이 필요
        - Network-level Search Space (전체 네트워크 탐색) : 
            - 	Depth dimension → 각 스테이지에 몇 개의 블록을 쌓을지
            - Width dimension → 각 레이어의 채널 수를 몇 개로 할지
            - Resolution dimension → 입력 이미지의 해상도
        - Cell-level이 네트워크를 구성하는 기본 단위(Cell)의 구조 자체를 탐색한다면, Network-level은 이미 cell이 정해져 있다고 가정하고, 그것을 어떻게 쌓고 연결하고 구성할지 (Depth, Width, Resolution 등)를 탐색한다고 이해함.
    2. Search Strategy : 그 공간 안에서 어떤 아키텍처를 탐색할지 결정하는 전략 (탐색 알고리즘)
        - how do we narrow down the design space?
        - **58페이지 Design space principle**
            - FLOPs(연산량) 분포를 기준으로 분석하는 방법 설명
            - 모델 학습 없이도 좋은 search space를 heuristic하게 고르는 방법: “FLOPs 분포”를 본다
            - 같은 memory constraint 하에서 더 많은 FLOPs를 수용할 수 있는 설계 공간이 좋은 설계 공간
                - Y축 Cumulative Probability : 특정 값 이하의 FLOPs를 가진 모델이, 전체 모델 중에서 얼마나 되는가?
                - 검정 곡선(w0.3-r160) : width = 0.3, resolution = 160
                    - 채널 수가 적고, 해상도는 높음
                    - 높은 해상도지만, 너비가 좁아 capacity가 떨어짐 (BAD)
                    -  x=32.3M일 때 y=80% → 검정 설계 공간의 모델 80%는 32.3M FLOPs 이하
                - 빨간 곡선(w0.5-r144) : width = 0.5, resolution = 144
                    - 해상도는 조금 낮지만, 채널 수가 더 많아짐
                    - 상대적으로 낮은 해상도지만, 너비가 넓어서 FLOPs가 많고 정확도도 높음 (GOOD)
                    -  x=50.3M일 때 y=80% → 빨간 설계 공간 내의 모델 중 80%가 50.3M FLOPs 이하
                    - 메모리 안에서 최대한의 연산” → 높은 정확도 기대 가능
                - 결국, 단순히 resolution만 높다고 좋은 게 아님 → FLOPs를 기준으로 판단해야 함
        - RL 기반 : NAS를 **순차적 결정 문제(sequential decision problem)**로 모델링, NN-based controller가 아키텍처를 순서대로 생성하면 -> 생성한 아키텍처로 child model을 학습시켜 accuracy / latency 등 보상 (reward) 측정 -> 이 보상 값을 사용해 controller를 업데이트
        - **Gradient Descent 기반** 이 부분 스터디 시간에 보충 필요
            - 기존의 방식이 여러 선택지에서 하나를 택하는 비연속적인 함수라면, 이걸 미분으로 풀기 위해 각 선택지에 각각 몇%의 가중치를 둘 것인지를 표현하는 연속적인 함수로 바꾸어 표현 -> 미분 가능하게 만들어서 search
            - latency를 loss함수에 penalty로 추가 -> latency를 줄이는 방향으로 학습 유도 가능

**Comment by tjdtnsu** [comment]
- bottleneck block, grouped convolution, depthwise-separable block, inverted bottleneck block, channel shuffle
- Multi-Head Self-Attention이 group별로 나누는 것이 아니라 Project하는 것인가?
- cell level search, network level search 할 수 있음. 이는 depth/resolution 등등 있을듯
- search strategy: grid, random, rainforcement, gradient, evolutionary search 있음
- 특히 gradient는 latency도 포함할 수 있음

**Comment by tjdtnsu** [comment]
![image](https://github.com/user-attachments/assets/ef0f8ebc-fad5-475a-9c00-3c68228851e5)
![image](https://github.com/user-attachments/assets/633629a7-f405-4ed6-858c-84fb089d6142)
![image](https://github.com/user-attachments/assets/e8eabcb3-ff42-491a-8b3f-e02b73593d14)
![image](https://github.com/user-attachments/assets/0d115361-19df-46cc-aa0c-be6c004d2910)
![image](https://github.com/user-attachments/assets/1134706f-b3fb-4601-b747-bd4d6cfbdaf7)
